// ng-lottie/__ivy_ngcc__/dist/esm/src/lottieAnimationView.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { LottieAnimationViewComponent } from './lottieAnimationView.component';
import * as ɵngcc0 from '@angular/core';
var LottieAnimationViewModule = (function () {
    function LottieAnimationViewModule() {
    }
    LottieAnimationViewModule.forRoot = function () {
        return {
            ngModule: LottieAnimationViewModule
        };
    };
LottieAnimationViewModule.ɵfac = function LottieAnimationViewModule_Factory(t) { return new (t || LottieAnimationViewModule)(); };
LottieAnimationViewModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: LottieAnimationViewModule });
LottieAnimationViewModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[CommonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(LottieAnimationViewModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    LottieAnimationViewComponent
                ],
                imports: [CommonModule],
                exports: [LottieAnimationViewComponent]
            }]
    }], function () { return []; }, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(LottieAnimationViewModule, { declarations: function () { return [LottieAnimationViewComponent]; }, imports: function () { return [CommonModule]; }, exports: function () { return [LottieAnimationViewComponent]; } }); })();
    return LottieAnimationViewModule;
}());
export { LottieAnimationViewModule };
/** @nocollapse */
LottieAnimationViewModule.ctorParameters = function () { return []; };
