// recorder-audio-worklet/build/es5/bundle.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@babel/runtime/helpers/defineProperty'), require('@babel/runtime/helpers/asyncToGenerator'), require('@babel/runtime/regenerator'), require('fast-unique-numbers'), require('worker-factory')) :
    typeof define === 'function' && define.amd ? define(['exports', '@babel/runtime/helpers/defineProperty', '@babel/runtime/helpers/asyncToGenerator', '@babel/runtime/regenerator', 'fast-unique-numbers', 'worker-factory'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.recorderAudioWorklet = {}, global._defineProperty, global._asyncToGenerator, global._regeneratorRuntime, global.fastUniqueNumbers, global.workerFactory));
}(this, (function (exports, _defineProperty, _asyncToGenerator, _regeneratorRuntime, fastUniqueNumbers, workerFactory) { 'use strict';

    function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

    var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);
    var _asyncToGenerator__default = /*#__PURE__*/_interopDefaultLegacy(_asyncToGenerator);
    var _regeneratorRuntime__default = /*#__PURE__*/_interopDefaultLegacy(_regeneratorRuntime);

    // This is the minified and stringified code of the recorder-audio-worklet-processor package.
    var worklet = "(()=>{var e={228:e=>{e.exports=function(e,t){(null==t||t>e.length)&&(t=e.length);for(var r=0,o=new Array(t);r<t;r++)o[r]=e[r];return o},e.exports.default=e.exports,e.exports.__esModule=!0},858:e=>{e.exports=function(e){if(Array.isArray(e))return e},e.exports.default=e.exports,e.exports.__esModule=!0},506:e=>{e.exports=function(e){if(void 0===e)throw new ReferenceError(\"this hasn't been initialised - super() hasn't been called\");return e},e.exports.default=e.exports,e.exports.__esModule=!0},575:e=>{e.exports=function(e,t){if(!(e instanceof t))throw new TypeError(\"Cannot call a class as a function\")},e.exports.default=e.exports,e.exports.__esModule=!0},100:(e,t,r)=>{var o=r(489),n=r(67);function s(t,r,u){return n()?(e.exports=s=Reflect.construct,e.exports.default=e.exports,e.exports.__esModule=!0):(e.exports=s=function(e,t,r){var n=[null];n.push.apply(n,t);var s=new(Function.bind.apply(e,n));return r&&o(s,r.prototype),s},e.exports.default=e.exports,e.exports.__esModule=!0),s.apply(null,arguments)}e.exports=s,e.exports.default=e.exports,e.exports.__esModule=!0},913:e=>{function t(e,t){for(var r=0;r<t.length;r++){var o=t[r];o.enumerable=o.enumerable||!1,o.configurable=!0,\"value\"in o&&(o.writable=!0),Object.defineProperty(e,o.key,o)}}e.exports=function(e,r,o){return r&&t(e.prototype,r),o&&t(e,o),e},e.exports.default=e.exports,e.exports.__esModule=!0},754:e=>{function t(r){return e.exports=t=Object.setPrototypeOf?Object.getPrototypeOf:function(e){return e.__proto__||Object.getPrototypeOf(e)},e.exports.default=e.exports,e.exports.__esModule=!0,t(r)}e.exports=t,e.exports.default=e.exports,e.exports.__esModule=!0},205:(e,t,r)=>{var o=r(489);e.exports=function(e,t){if(\"function\"!=typeof t&&null!==t)throw new TypeError(\"Super expression must either be null or a function\");e.prototype=Object.create(t&&t.prototype,{constructor:{value:e,writable:!0,configurable:!0}}),t&&o(e,t)},e.exports.default=e.exports,e.exports.__esModule=!0},430:e=>{e.exports=function(e){return-1!==Function.toString.call(e).indexOf(\"[native code]\")},e.exports.default=e.exports,e.exports.__esModule=!0},67:e=>{e.exports=function(){if(\"undefined\"==typeof Reflect||!Reflect.construct)return!1;if(Reflect.construct.sham)return!1;if(\"function\"==typeof Proxy)return!0;try{return Boolean.prototype.valueOf.call(Reflect.construct(Boolean,[],(function(){}))),!0}catch(e){return!1}},e.exports.default=e.exports,e.exports.__esModule=!0},884:e=>{e.exports=function(e,t){var r=null==e?null:\"undefined\"!=typeof Symbol&&e[Symbol.iterator]||e[\"@@iterator\"];if(null!=r){var o,n,s=[],u=!0,p=!1;try{for(r=r.call(e);!(u=(o=r.next()).done)&&(s.push(o.value),!t||s.length!==t);u=!0);}catch(e){p=!0,n=e}finally{try{u||null==r.return||r.return()}finally{if(p)throw n}}return s}},e.exports.default=e.exports,e.exports.__esModule=!0},521:e=>{e.exports=function(){throw new TypeError(\"Invalid attempt to destructure non-iterable instance.\\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.\")},e.exports.default=e.exports,e.exports.__esModule=!0},585:(e,t,r)=>{var o=r(8).default,n=r(506);e.exports=function(e,t){return!t||\"object\"!==o(t)&&\"function\"!=typeof t?n(e):t},e.exports.default=e.exports,e.exports.__esModule=!0},489:e=>{function t(r,o){return e.exports=t=Object.setPrototypeOf||function(e,t){return e.__proto__=t,e},e.exports.default=e.exports,e.exports.__esModule=!0,t(r,o)}e.exports=t,e.exports.default=e.exports,e.exports.__esModule=!0},38:(e,t,r)=>{var o=r(858),n=r(884),s=r(379),u=r(521);e.exports=function(e,t){return o(e)||n(e,t)||s(e,t)||u()},e.exports.default=e.exports,e.exports.__esModule=!0},8:e=>{function t(r){return\"function\"==typeof Symbol&&\"symbol\"==typeof Symbol.iterator?(e.exports=t=function(e){return typeof e},e.exports.default=e.exports,e.exports.__esModule=!0):(e.exports=t=function(e){return e&&\"function\"==typeof Symbol&&e.constructor===Symbol&&e!==Symbol.prototype?\"symbol\":typeof e},e.exports.default=e.exports,e.exports.__esModule=!0),t(r)}e.exports=t,e.exports.default=e.exports,e.exports.__esModule=!0},379:(e,t,r)=>{var o=r(228);e.exports=function(e,t){if(e){if(\"string\"==typeof e)return o(e,t);var r=Object.prototype.toString.call(e).slice(8,-1);return\"Object\"===r&&e.constructor&&(r=e.constructor.name),\"Map\"===r||\"Set\"===r?Array.from(e):\"Arguments\"===r||/^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(r)?o(e,t):void 0}},e.exports.default=e.exports,e.exports.__esModule=!0},957:(e,t,r)=>{var o=r(754),n=r(489),s=r(430),u=r(100);function p(t){var r=\"function\"==typeof Map?new Map:void 0;return e.exports=p=function(e){if(null===e||!s(e))return e;if(\"function\"!=typeof e)throw new TypeError(\"Super expression must either be null or a function\");if(void 0!==r){if(r.has(e))return r.get(e);r.set(e,t)}function t(){return u(e,arguments,o(this).constructor)}return t.prototype=Object.create(e.prototype,{constructor:{value:t,enumerable:!1,writable:!0,configurable:!0}}),n(t,e)},e.exports.default=e.exports,e.exports.__esModule=!0,p(t)}e.exports=p,e.exports.default=e.exports,e.exports.__esModule=!0},242:function(e,t,r){!function(e,t,r,o,n,s,u){\"use strict\";function p(e){return e&&\"object\"==typeof e&&\"default\"in e?e:{default:e}}var a=p(e),l=p(t),i=p(r),c=p(o),f=p(n),d=p(s);function x(e){var t=_();return function(){var r,o=d.default(e);if(t){var n=d.default(this).constructor;r=Reflect.construct(o,arguments,n)}else r=o.apply(this,arguments);return f.default(this,r)}}function _(){if(\"undefined\"==typeof Reflect||!Reflect.construct)return!1;if(Reflect.construct.sham)return!1;if(\"function\"==typeof Proxy)return!0;try{return Boolean.prototype.valueOf.call(Reflect.construct(Boolean,[],(function(){}))),!0}catch(e){return!1}}var y=function(e){c.default(r,e);var t=x(r);function r(){var e;return l.default(this,r),(e=t.call(this))._encoderPort=null,e._state=\"inactive\",e.port.onmessage=function(t){var r=t.data;\"record\"===r.method?\"inactive\"===e._state?(e._encoderPort=r.params.encoderPort,e._state=\"active\",e.port.postMessage({id:r.id,result:null})):e.port.postMessage({error:{code:-32603,message:\"The internal state does not allow to process the given message.\"},id:r.id}):\"stop\"===r.method?\"active\"!==e._state&&\"recording\"!==e._state||null===e._encoderPort?e.port.postMessage({error:{code:-32603,message:\"The internal state does not allow to process the given message.\"},id:r.id}):(e._stop(e._encoderPort),e.port.postMessage({id:r.id,result:null})):\"number\"==typeof r.id&&e.port.postMessage({error:{code:-32601,message:\"The requested method is not supported.\"},id:r.id})},e}return i.default(r,[{key:\"process\",value:function(e){var t=a.default(e,1)[0];if(\"inactive\"===this._state)return!0;if(\"active\"===this._state){if(void 0===t)throw new Error(\"No channelData was received for the first input.\");if(0===t.length)return!0;this._state=\"recording\"}if(\"recording\"===this._state&&null!==this._encoderPort){if(void 0===t)throw new Error(\"No channelData was received for the first input.\");if(0!==t.length)return this._encoderPort.postMessage(t,t.map((function(e){return e.buffer}))),!0;this._stop(this._encoderPort)}return!1}},{key:\"_stop\",value:function(e){e.postMessage([]),e.close(),this._encoderPort=null,this._state=\"stopped\"}}]),r}(p(u).default(AudioWorkletProcessor));y.parameterDescriptors=[],registerProcessor(\"recorder-audio-worklet-processor\",y)}(r(38),r(575),r(913),r(205),r(585),r(754),r(957))}},t={};function r(o){var n=t[o];if(void 0!==n)return n.exports;var s=t[o]={exports:{}};return e[o].call(s.exports,s,s.exports,r),s.exports}r.n=e=>{var t=e&&e.__esModule?()=>e.default:()=>e;return r.d(t,{a:t}),t},r.d=(e,t)=>{for(var o in t)r.o(t,o)&&!r.o(e,o)&&Object.defineProperty(e,o,{enumerable:!0,get:t[o]})},r.o=(e,t)=>Object.prototype.hasOwnProperty.call(e,t),(()=>{\"use strict\";r(242)})()})();"; // tslint:disable-line:max-line-length

    function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

    function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty__default['default'](target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }
    var blob = new Blob([worklet], {
      type: 'application/javascript; charset=utf-8'
    });
    var addRecorderAudioWorkletModule = /*#__PURE__*/function () {
      var _ref = _asyncToGenerator__default['default']( /*#__PURE__*/_regeneratorRuntime__default['default'].mark(function _callee(addAudioWorkletModule) {
        var url;
        return _regeneratorRuntime__default['default'].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                url = URL.createObjectURL(blob);
                _context.prev = 1;
                _context.next = 4;
                return addAudioWorkletModule(url);

              case 4:
                _context.prev = 4;
                URL.revokeObjectURL(url);
                return _context.finish(4);

              case 7:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, null, [[1,, 4, 7]]);
      }));

      return function addRecorderAudioWorkletModule(_x) {
        return _ref.apply(this, arguments);
      };
    }();
    function createRecorderAudioWorkletNode(audioWorkletNodeConstructor, context) {
      var options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
      var audioWorkletNode = new audioWorkletNodeConstructor(context, 'recorder-audio-worklet-processor', _objectSpread(_objectSpread({}, options), {}, {
        channelCountMode: 'explicit',
        numberOfInputs: 1,
        numberOfOutputs: 0
      }));
      var ongoingRequests = new Map();

      var listener = function listener(_ref2) {
        var message = _ref2.data;
        var id = message.id;

        if (id !== null && ongoingRequests.has(id)) {
          var _ongoingRequests$get = ongoingRequests.get(id),
              reject = _ongoingRequests$get.reject,
              resolve = _ongoingRequests$get.resolve;

          ongoingRequests["delete"](id);

          if (message.error === undefined) {
            resolve(message.result);
          } else {
            reject(new Error(message.error.message));
          }
        }
      };

      var postMessage = function (port) {
        return function (message) {
          var transferables = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : [];
          return new Promise(function (resolve, reject) {
            var id = fastUniqueNumbers.generateUniqueNumber(ongoingRequests);
            ongoingRequests.set(id, {
              reject: reject,
              resolve: resolve
            });
            port.postMessage(_objectSpread({
              id: id
            }, message), transferables);
          });
        };
      }(audioWorkletNode.port);

      var removeEventListener = function (port) {
        port.addEventListener('message', listener);
        port.start();
        return function () {
          return port.removeEventListener('message', listener);
        };
      }(audioWorkletNode.port);

      var state = 'inactive';

      var changeState = function changeState(expectedState, nextState) {
        if (state !== expectedState) {
          throw new Error("Expected the state to be \"".concat(expectedState, "\" but it was \"").concat(state, "\"."));
        }

        state = nextState;
      };

      Object.defineProperties(audioWorkletNode, {
        port: {
          get: function get() {
            throw new Error("The port of a RecorderAudioWorkletNode can't be accessed.");
          }
        },
        record: {
          get: function get() {
            return /*#__PURE__*/function () {
              var _ref3 = _asyncToGenerator__default['default']( /*#__PURE__*/_regeneratorRuntime__default['default'].mark(function _callee2(encoderPort) {
                return _regeneratorRuntime__default['default'].wrap(function _callee2$(_context2) {
                  while (1) {
                    switch (_context2.prev = _context2.next) {
                      case 0:
                        changeState('inactive', 'recording');
                        return _context2.abrupt("return", postMessage({
                          method: 'record',
                          params: {
                            encoderPort: encoderPort
                          }
                        }, [encoderPort]));

                      case 2:
                      case "end":
                        return _context2.stop();
                    }
                  }
                }, _callee2);
              }));

              return function (_x2) {
                return _ref3.apply(this, arguments);
              };
            }();
          }
        },
        stop: {
          get: function get() {
            return /*#__PURE__*/_asyncToGenerator__default['default']( /*#__PURE__*/_regeneratorRuntime__default['default'].mark(function _callee3() {
              return _regeneratorRuntime__default['default'].wrap(function _callee3$(_context3) {
                while (1) {
                  switch (_context3.prev = _context3.next) {
                    case 0:
                      changeState('recording', 'stopped');
                      _context3.prev = 1;
                      _context3.next = 4;
                      return postMessage({
                        method: 'stop'
                      });

                    case 4:
                      _context3.prev = 4;
                      removeEventListener();
                      return _context3.finish(4);

                    case 7:
                    case "end":
                      return _context3.stop();
                  }
                }
              }, _callee3, null, [[1,, 4, 7]]);
            }));
          }
        }
      });
      return audioWorkletNode;
    }

    Object.defineProperty(exports, 'isSupported', {
        enumerable: true,
        get: function () {
            return workerFactory.isSupported;
        }
    });
    exports.addRecorderAudioWorkletModule = addRecorderAudioWorkletModule;
    exports.createRecorderAudioWorkletNode = createRecorderAudioWorkletNode;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
